vms/platformvm/validators_set_property_test.go

```diff
@@ -0,0 +1,854 @@
+	"github.com/ava-labs/avalanchego/database/prefixdb"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/uptime"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/crypto/secp256k1"
+	"github.com/ava-labs/avalanchego/utils/formatting"
+	"github.com/ava-labs/avalanchego/utils/formatting/address"
+	"github.com/ava-labs/avalanchego/utils/json"
+	"github.com/ava-labs/avalanchego/utils/timer/mockable"
+	"github.com/ava-labs/avalanchego/utils/units"
+	"github.com/ava-labs/avalanchego/version"
+	"github.com/ava-labs/avalanchego/vms/avm/blocks/executor"
+	"github.com/ava-labs/avalanchego/vms/components/avax"
+	"github.com/ava-labs/avalanchego/vms/platformvm/api"
+	"github.com/ava-labs/avalanchego/vms/platformvm/blocks"
+	"github.com/ava-labs/avalanchego/vms/platformvm/config"
+	"github.com/ava-labs/avalanchego/vms/platformvm/reward"
+	"github.com/ava-labs/avalanchego/vms/platformvm/signer"
+	"github.com/ava-labs/avalanchego/vms/platformvm/state"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	"github.com/ava-labs/avalanchego/vms/platformvm/utxo"
+	"github.com/ava-labs/avalanchego/vms/secp256k1fx"
+
+	blockexecutor "github.com/ava-labs/avalanchego/vms/platformvm/blocks/executor"
+)
+
+const (
+	startPrimaryWithBLS    = "primary-validator-with-BLS-key\n"
+	startPrimaryWithoutBLS = "primary-validator-without-BLS-key\n"
+	startSubnetValidator   = "subnet-validator\n"
+)
+
+var errEmptyEventsList = errors.New("empty events list")
+
+// for a given (permissioned) subnet, the test stakes and restakes multiple times
+// a node as a primary and subnet validator. The BLS key of the node is changed across
+// staking periods, and it can even be nil. We test that GetValidatorSet returns
+// the correct primary and subnet validators data, with the right BLS key version at
+// all relevant heights.
+func TestGetValidatorsSetProperty(t *testing.T) {
+	properties := gopter.NewProperties(nil)
+
+	// to reproduce a given scenario do something like this:
+	// parameters := gopter.DefaultTestParametersWithSeed(1685887576153675816)
+	// properties := gopter.NewProperties(parameters)
+
+	properties.Property("check GetValidatorSet", prop.ForAll(
+		func(events []string) string {
+			vm, subnetID, err := buildVM()
+			if err != nil {
+				return fmt.Sprintf("failed building vm, %s", err.Error())
+			}
+			vm.ctx.Lock.Lock()
+			defer func() {
+				_ = vm.Shutdown(context.Background())
+				vm.ctx.Lock.Unlock()
+			}()
+			nodeID := ids.GenerateTestNodeID()
+
+			currentTime := defaultGenesisTime
+			vm.clock.Set(currentTime)
+			vm.state.SetTimestamp(currentTime)
+
+			// build validator sequences out of pseudo-random events
```

Comment:
can we clarify this comment a little. is this validator transitions?

Reply:
hopefully done